### PR TITLE
[Feature 6] Filtering Functionality

### LINKED ISSUE #11: [Feature 6] Filtering Functionality
* method for filtering users by certain criteria
* updating the view to display the filtered users

### PR DESCRIPTION
**Pull request by helen-li**
Fixes #11 

Example filtering UI:

Implementation details:
* Created `activity_academic_filter.xml` with all the UI components and layout for filtering feed
* Created `AcademicFilterActivity` class as the View corresponding to the filter functionality
* Integrated use case method for the filter functionality (`filterCompatibilityList`) into the main use case interactor for feed-related functionalities (`GenerateCompatibilityList.java`). The method filters through a list of compatible users (of the current user) based on certain given criteria for filters (provided via user input through the UI)
* Updated `RecommendationPresenter` with conditionals to display filtered list of compatible users when filters are selected while maintaining the default behavior (displaying the list of compatible users when filters are not selected)
* Created units tests with mock Users and filters in `FilterUseCaseUnitTest.java`

Refactoring and other small bug fixes:
* Grouped all classes related to the Recommendation Feed into one package
* Started putting `strings.xml` to use — as a next step, need to refactor and store all constants used in xml files in here (created new Issue #33)
* Updated minSdk version from 23 to 24 to allow access to sorting methods in `Comparator` (bug fix for #22)
* Sporadic changes throughout the repo from `ArrayList` to `List` and `HashSet` to `Set` whenever possible/applicable
* Removed unnecessary Presenter Interface since Presenter layer can call Use Case layer directly (bug fix for #23)
* Added mock user Id arguments to pre-existing test cases that tried to create mock users. This was causing errors because #24 made changes to the User constructor (added a parameter)

### CODE REVIEW

**Review by MilenaMijuskovic — DISMISSED** [review]
Nice comments! Your code looks good :). Your UI also looks really nice. I think the final step now is once we all merge our code together to make sure we are flipping through the views correctly (i.e when a user is on the recommendation feed page and wants to go to the filter page, that the views switch without any errors). Other than that, looks great!

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (nicholaspoon03) [inline app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:12]: I don't know if I'm missing something but where exactly does this class get the compatibility list. It seems that you've initialized the compatibilityList without actually giving it a value anywhere? I presume that this will be later initialized in the constructor as it is empty at the moment but I just want to make sure. Also, I want to make a note to myself that we need to discuss how to store the users in the compatibilityList. Should it be a list of User objects or a list of strings (Uids). It is probably better to have it a list of User objects since most of our methods and classes require the actual user object but we should probably talk about this and have a concrete plan.

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (nicholaspoon03) [inline app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:36]: I just wanted to make sure that since we are mutating the compatibilityList that the actual compatibilityList is not modified (but only the compatibilityList that is defined in this class). I don't really know how the behaviour will work since I don't see the initializing (or retireval) of the compatibilityList.

**Review by nicholaspoon03 — COMMENTED** [review]
Good job on everything, Helen! Documentation is really well written and clearly explains what is being done in each method and class. The comments are also really helpful to follow along with your logic. Just a few concerns that I pointed out but those should be resolved during our next meeting. Your UI is also really nice and well laid out, as Milena has pointed out.

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:12]: Thanks Nicholas! I'm waiting on your PR to be merged to add getters and setters into your class to grab compatibility list. If you scroll down further in this same class, there are two methods commented out. They correspond to what I briefly messaged you offline about regarding using one method to turn a list of String Uids into a list of User objects. It might be more efficient for us to chat the details of integrating these offline.

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:36]: Good question! My answer is related to the thread right above this one - I think everything will become much more clear when your PR #22 gets merged into main so you can see my changes in context of your `GenerateCompatibilityList` to understand how I was intending to structure the filtering functionality to work :)

File: app/src/test/java/com/group80/uoftinder/RecommendationInteractorUnitTest.java
Comment (rsrinivasan1) [inline app/src/test/java/com/group80/uoftinder/RecommendationInteractorUnitTest.java:17]: Test cases look great! Nice work on covering all major cases with filtering.

**Review by rsrinivasan1 — COMMENTED** [review]
Great work all around, Helen! Your documentation is very clear throughout, and your code was easy enough to understand for someone who hasn't looked at activities and UI elements in Android. As you mentioned in one of the threads with Nicholas, I'm sure the behavior of some of these methods will become more clear when other PRs get merged.

Regarding ```filterCompatibilityList```, I was having some trouble understanding exactly how the i-th element of ```filters``` and ```answers``` relate to each other, so maybe you could add some small comments on the functionality beside those lines. 

Otherwise, test cases look great, and I'm happy to see the filter UI is working!

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:36]: Yup for sure! When a user decides to filter their feed, they expect to see recommended users who match the filtering criteria they provided. The way we currently show recommended users to them is via compatibility list where the top element gets passed to the Presenter and then the View. In order to show filtered feed, we must filter compatibility list which requires changing the list. Nicholas' comment did raise a reminder to me that maybe we want to preserve the original list in case users want to reset their filters at any time or change their filters - I have a solution in mind for how to approach this, just need to be able to access the code in Nicholas' PR once it is merged. Hope this clarifies things up slightly :)

File: app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java
Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java:116]: Ah good catch! This thought did cross my mind at some point - just wasn't a priority in this PR. Definitely not a nitpick since it could very much happen to users - I think it would be the easiest for the View to catch this mismatch error via form validation. Just created a new issue #29 to reflect this as a next step.

Comment (rsrinivasan1) [reply app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java:116]: Sounds good!

File: app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java
Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java:124]: 100% correct on that. It should go back to whichever View will be displaying the main Recommendation Feed (hence the comment above that starts with TODO). Not sure what this class will be called at the moment. The vision is that I will add a button to that View. When this button gets clicked on, the Filter UI opens. When the user is done choosing filters, they should be redirected to that View to see the filtered feed.

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:42]: This code operates on the assumption that the number of sets within answers and filters will always be the same (`filters.size() == answers.size()`), which should be true since the filters are created based on the exact questions that users have to answer to build their profile.

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (rsrinivasan1) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:36]: Yep, thanks for the explanation! I was wondering about how you were planning to preserve the original list in case the filter changes, but sounds like you should have that covered.

File: app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java
Comment (rsrinivasan1) [reply app/src/main/java/com/group80/uoftinder/AcademicFilterActivity.java:124]: Ah I see, I totally missed the point of that TODO comment.

File: app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java
Comment (rsrinivasan1) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:42]: Okay, I see. Maybe you could specifically comment that in your code as an invariant. I'm a bit concerned that as the number of questions gets larger (for instance, if we include 10 questions like we were planning to at one point), there may not be enough space on screen to display every option for each question. Have you considered what will happen at that stage?

Comment (helen-li) [reply app/src/main/java/com/group80/uoftinder/RecommendationInteractor.java:44]: Very valid concern, I considered that as well. My solution was with the UI where I implemented a scrollable view so that when the number of questions increase by a lot, the filters would still fit "on" the screen due to its scrollable nature. Just added more comments about the invariant in my most recent commit!

**Review by MilenaMijuskovic — APPROVED** [review]
Nice!

**Review by nicholaspoon03 — APPROVED** [review]
Looks good, Helen! Thanks for the hard work! Your changes to GenerateCompatibilityList make a lot of sense to me and seem correct!

**Review by rsrinivasan1 — APPROVED** [review]
Thanks for figuring out how all of this fits together, Helen! Looks like the filtering functionality interacts as intended with Nick's code for the "generate compatibility list" use case. 

Excited to see how the final app will look!

**Pull request merged**
